src/core/blocks/casting.cpp

```diff
@@ -242,278 +242,113 @@ template <CBType FROMTYPE> struct ToBytes {
-      case Float:                                                              \
-      case Float2:                                                             \
-      case Float3:                                                             \
-      case Float4:                                                             \
-      case String:                                                             \
-        if (convert(output, input))                                            \
-          return output;                                                       \
-        break;                                                                 \
-      default:                                                                 \
-        throw ActivationError("Cannot cast given input type.");                \
-      }                                                                        \
-      return output;                                                           \
-    }                                                                          \
-  };                                                                           \
-  RUNTIME_CORE_BLOCK(To##_varName_);                                           \
-  RUNTIME_BLOCK_inputTypes(To##_varName_);                                     \
-  RUNTIME_BLOCK_outputTypes(To##_varName_);                                    \
-  RUNTIME_BLOCK_compose(To##_varName_);                                        \
-  RUNTIME_BLOCK_activate(To##_varName_);                                       \
-  RUNTIME_BLOCK_END(To##_varName_);
-
-TO_SOMETHING_SIMPLE(Int, int64_t, intValue, stoll, IntType);
-TO_SOMETHING(Int, 2, int64_t, int2Value, stoll, Int2Type);
-TO_SOMETHING(Int, 3, int32_t, int3Value, stol, Int3Type);
-TO_SOMETHING(Int, 4, int32_t, int4Value, stol, Int4Type);
-TO_SOMETHING_SIMPLE(Float, double, floatValue, stod, FloatType);
-TO_SOMETHING(Float, 2, double, float2Value, stod, Float2Type);
-TO_SOMETHING(Float, 3, float, float3Value, stof, Float3Type);
-TO_SOMETHING(Float, 4, float, float4Value, stof, Float4Type);
+template <CBType ToType> struct ToNumber {
+  const VectorTypeTraits *_outputVectorType{nullptr};
+  const NumberTypeTraits *_outputNumberType{nullptr};
+
+  CBTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  CBTypesInfo outputTypes() {
+    if(!_outputVectorType)
+      return CoreInfo::AnyType;
+    return _outputVectorType->type; 
+  }
+  
+  CBTypeInfo compose(const CBInstanceData &data) {
+    _outputVectorType =
+        VectorTypeLookup::getInstance().get(ToType);
+    if (!_outputVectorType) {
+      throw ComposeError("Conversion not implemented for this type");
+    }
+
+    _outputNumberType = NumberTypeLookup::getInstance().get(_outputVectorType->numberType);
+
+    return _outputVectorType->type;
+  }
+
+  void parseVector(CBVar &output, const CBVar &input) {
+    const VectorTypeTraits *inputVectorType =
```

Comment:
Input type won't change, can we move those lookups into `warmup` and cache them?